diff --git a/linux-user/syscall.c b/linux-user/syscall.c
index b13a170e..4af79175 100644
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -11504,6 +11504,28 @@
         return do_swapcontext(cpu_env, arg1, arg2, arg3);
 #endif

+    /* Custom syscall for fuzzing JSC JIT */
+    case 536:
+        printf("[QEMU] Caught Syscall 536\n");
+        int afl_input;
+        FILE *afl_input_fptr;
+        if ((afl_input_fptr = fopen("/tmp/.cur_input","r")) == NULL) {
+            printf("Error opening file!");
+            return -1;
+        }
+        fscanf(afl_input_fptr,"%d", &afl_input);
+        printf("[QEMU] AFL input: %d\n", afl_input);
+        // Read contents from file
+        printf("[QEMU] AFL input (hex encoded): 0x");
+        char afl_input_char = fgetc(afl_input_fptr);
+        while (afl_input_char != EOF) {
+            printf("%x", afl_input_char);
+            afl_input_char = fgetc(afl_input_fptr);
+        }
+        fclose(afl_input_fptr);
+        printf("\n");
+        return afl_input;
+
     default:
         qemu_log_mask(LOG_UNIMP, "Unsupported syscall: %d\n", num);
         return -TARGET_ENOSYS;
