diff --git a/Sources/Fuzzilli/Fuzzer.swift b/Sources/Fuzzilli/Fuzzer.swift
index 632efb6..e07637e 100644
--- a/Sources/Fuzzilli/Fuzzer.swift
+++ b/Sources/Fuzzilli/Fuzzer.swift
@@ -601,15 +601,15 @@ public class Fuzzer {
         }
 
         #if os(Linux)
         do {
             let corePattern = try String(contentsOfFile: "/proc/sys/kernel/core_pattern", encoding: String.Encoding.ascii)
-            if !corePattern.hasPrefix("|/bin/false") {
+            if !corePattern.hasPrefix("|/bin/false") && !corePattern.hasPrefix("core") {
                 logger.fatal("Please run: sudo sysctl -w 'kernel.core_pattern=|/bin/false'")
             }
         } catch {
-            logger.warning("Could not check core dump behaviour. Please ensure core_pattern is set to '|/bin/false'")
+            logger.warning("Could not check core dump behaviour. Please ensure core_pattern is set to '|/bin/false' or 'core'")
         }
         #endif
 
         // Check if we can execute programs
         var execution = execute(Program())
